pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,479 @@
+	} else {
+		err = bfs.fetchOIDsFromIndexFiles()
+	}
+	var force bool
+	if err != nil {
+		bfs.log.Error("NeoFS BlockFetcher service: OID downloading routine failed", zap.Error(err))
+		force = true
+	}
+	// Stop the service since there's nothing to do anymore.
+	bfs.stopService(force)
+}
+
+// blockDownloader downloads the block from NeoFS and sends it to the blocks channel.
+func (bfs *Service) blockDownloader() {
+	defer bfs.wg.Done()
+
+	for blkOid := range bfs.oidsCh {
+		ctx, cancel := context.WithTimeout(bfs.ctx, bfs.cfg.Timeout)
+		defer cancel()
+
+		rc, err := bfs.objectGet(ctx, blkOid.String())
+		if err != nil {
+			if isContextCanceledErr(err) {
+				return
+			}
+			bfs.log.Error("failed to objectGet block", zap.Error(err))
+			bfs.stopService(true)
+			return
+		}
+
+		b, err := bfs.readBlock(rc)
+		if err != nil {
+			if isContextCanceledErr(err) {
+				return
+			}
+			bfs.log.Error("failed to read block", zap.Error(err))
+			bfs.stopService(true)
+			return
+		}
+		select {
+		case <-bfs.ctx.Done():
+			return
+		case bfs.blocksCh <- b:
+		}
+	}
+}
+
+// blockQueuer puts the block into the bqueue.
+func (bfs *Service) blockQueuer() {
+	defer close(bfs.blockQueuerToExiter)
+
+	for b := range bfs.blocksCh {
+		select {
+		case <-bfs.ctx.Done():
+			return
+		default:
+			err := bfs.enqueueBlock(b)
+			if err != nil {
+				bfs.log.Error("failed to enqueue block", zap.Error(err))
+				bfs.stopService(true)
+				return
+			}
+		}
+	}
+}
+
+// fetchOIDsFromIndexFiles fetches block OIDs from NeoFS by searching index files first.
+func (bfs *Service) fetchOIDsFromIndexFiles() error {
+	h := bfs.chain.BlockHeight()
+	startIndex := h/bfs.cfg.IndexFileSize + 1
+	skip := h % bfs.cfg.IndexFileSize
+
+	for {
+		select {
+		case <-bfs.exiterToOIDDownloader:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.IndexFileAttribute, fmt.Sprintf("%d", startIndex), object.MatchStringEqual)
+			prm.SetFilters(filters)
+
+			ctx, cancel := context.WithTimeout(bfs.ctx, bfs.cfg.Timeout)
+			blockOidsObject, err := bfs.objectSearch(ctx, prm)
+			cancel()
+			if err != nil {
+				if isContextCanceledErr(err) {
+					return nil
+				}
+				return fmt.Errorf("failed to find '%s' object with index %d: %w", bfs.cfg.IndexFileAttribute, startIndex, err)
+			}
+			if len(blockOidsObject) == 0 {
+				bfs.log.Info(fmt.Sprintf("NeoFS BlockFetcher service: no '%s' object found with index %d, stopping", bfs.cfg.IndexFileAttribute, startIndex))
+				return nil
+			}
+
+			blockCtx, blockCancel := context.WithTimeout(bfs.ctx, bfs.cfg.Timeout)
+			defer blockCancel()
+			oidsRC, err := bfs.objectGet(blockCtx, blockOidsObject[0].String())
+			if err != nil {
+				if isContextCanceledErr(err) {
+					return nil
+				}
+				return fmt.Errorf("failed to fetch '%s' object with index %d: %w", bfs.cfg.IndexFileAttribute, startIndex, err)
+			}
+
+			err = bfs.streamBlockOIDs(oidsRC, int(skip))
```

Comment:
One possible problem that I see with current design is that when we're getting OIDs from index files, a connection to storage node remains opened. We read OID files step-by-step and if destination channel is full then `streamBlockOIDs` will wait with opened connection to NeoFS storage node. Timeout may happen here.

Reply:
retry probably will help?

Reply:
It's a design problem.

Reply:
Ref. https://github.com/nspcc-dev/neo-go/issues/3585.